File: source/jugglinglab/core/ApplicationPanel.java
// ApplicationPanel.java
//
// Copyright 2002-2021 Jack Boyce and the Juggling Lab contributors

package jugglinglab.core;

import java.awt.*;
import java.awt.event.*;
import java.util.ResourceBundle;
import java.text.MessageFormat;
import javax.swing.*;
import javax.swing.event.*;

import jugglinglab.generator.*;
import jugglinglab.jml.JMLPattern;
import jugglinglab.notation.*;
import jugglinglab.util.*;
import jugglinglab.view.View;


// This class represents the entire contents of the ApplicationWindow frame.
// For a given notation type it creates a tabbed pane with separate panels for
// pattern entry, transitions, and generator.

public class ApplicationPanel extends JPanel implements ActionListener {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    protected JFrame parent;
    protected JTabbedPane jtp;
    protected View animtarget;
    protected PatternListPanel patlist;
    protected boolean patlisttab = false;

    protected int currentnum = -1;
    protected JButton juggle_button;
    protected JButton trans_button;
    protected JButton gen_button;
    protected JLabel gen_busy;

    // Execution limits for generator
    protected static final int max_patterns = 1000;
    protected static final double max_time = 15.0;


    public ApplicationPanel(JFrame parent) {
        this(parent, null, null, false);
    }

    public ApplicationPanel(JFrame p, View v, PatternListPanel pl, boolean patlisttab) {
        parent = p;

        // fields below are currently unused; they supported the applet version
        // of Juggling Lab
        animtarget = v;
        patlist = pl;
        patlisttab = (pl == null) ? patlisttab : true;
    }

    @Override
    public void actionPerformed(ActionEvent ae) {
        String command = ae.getActionCommand();

        try {
            if (command.startsWith("notation")) {
                try {
                    int num = Integer.parseInt(command.substring(8));
                    if (num != currentnum) {
                        setNotation(num);
                        if (parent != null)
                            parent.pack();
                        currentnum = num;
                    }
                } catch (NumberFormatException nfe) {
                    throw new JuggleExceptionInternal("Error in notation number coding");
                }
            }
        } catch (Exception e) {
            ErrorDialog.handleFatalException(e);
        }
    }

    // input is for example Pattern.NOTATION_SITESWAP
    public void setNotation(int num) throws JuggleExceptionUser, JuggleExceptionInternal {
        if (num > Pattern.builtinNotations.length)
            return;

        if (jtp != null)
            remove(jtp);
        jtp = new JTabbedPane();

        switch (num) {
            case Pattern.NOTATION_SITESWAP:
                addPatternEntryControl(new SiteswapNotationControl());
                break;
        }

        PatternListPanel pl = patlist;
        if (pl == null && patlisttab)
            pl = new PatternListPanel(animtarget);

        Transitioner trans = Transitioner.newTransitioner(Pattern.builtinNotations[num - 1]);
        if (trans != null)
            addTransitionerControl(trans, pl);

        Generator gen = Generator.newGenerator(Pattern.builtinNotations[num - 1]);
        if (gen != null)
            addGeneratorControl(gen, pl);

        if (pl != null) {
            jtp.addTab(guistrings.getString("Pattern_list_tab"), pl);
            if (patlist != null)
                jtp.setSelectedComponent(pl);  // if we loaded from a file
        }

        // Change the default button when the tab changes
        jtp.addChangeListener(new ChangeListener() {
            @Override
            public void stateChanged(ChangeEvent e) {
                getRootPane().setDefaultButton(getDefaultButton());
            }
        });

        setLayout(new BorderLayout());
        add(jtp, BorderLayout.CENTER);

        if (parent != null)
            parent.getRootPane().setDefaultButton(getDefaultButton());
    }

    protected void addPatternEntryControl(NotationControl control) {
        final NotationControl fcontrol = control;

        JPanel np1 = new JPanel();
        np1.setLayout(new BorderLayout());
        np1.add(control, BorderLayout.PAGE_START);

        JPanel np2 = new JPanel();
        np2.setLayout(new FlowLayout(FlowLayout.TRAILING));
        JButton nbut1 = new JButton(guistrings.getString("Defaults"));
        nbut1.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                try {
                    fcontrol.resetControl();
                } catch (Exception e) {
                    ErrorDialog.handleFatalException(e);
                }
            }
        });
        np2.add(nbut1);

        juggle_button = new JButton(guistrings.getString("Juggle"));
        juggle_button.setDefaultCapable(true);

        juggle_button.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                PatternWindow jaw2 = null;
                try {
                    ParameterList pl = fcontrol.getParameterList();
                    Pattern p = fcontrol.newPattern().fromParameters(pl);
                    AnimationPrefs jc = (new AnimationPrefs()).fromParameters(pl);
                    pl.errorIfParametersLeft();

                    String notation = p.getNotationName();
                    String config = p.toString();
                    JMLPattern pat = JMLPattern.fromBasePattern(notation, config);
                    pat.layoutPattern();

                    if (PatternWindow.bringToFront(pat.getHashCode()))
                        return;

                    if (animtarget != null)
                        animtarget.restartView(pat, jc);
                    else
                        jaw2 = new PatternWindow(pat.getTitle(), pat, jc);
                } catch (JuggleExceptionUser je) {
                    if (jaw2 != null)
                        jaw2.dispose();
                    new ErrorDialog(ApplicationPanel.this, je.getMessage());
                } catch (Exception e) {
                    if (jaw2 != null)
                        jaw2.dispose();
                    ErrorDialog.handleFatalException(e);
                }
            }
        });
        np2.add(juggle_button);
        np1.add(np2, BorderLayout.PAGE_END);

        jtp.addTab(guistrings.getString("Pattern_entry"), np1);
    }

    protected void addTransitionerControl(Transitioner trans, PatternListPanel pl) {
        JPanel p1 = new JPanel();
        p1.setLayout(new BorderLayout());
        p1.add(trans.getTransitionerControl(), BorderLayout.PAGE_START);
        JPanel p2 = new JPanel();
        p2.setLayout(new FlowLayout(FlowLayout.TRAILING));

        JButton but1 = new JButton(guistrings.getString("Defaults"));

        but1.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                trans.resetTransitionerControl();
            }
        });
        p2.add(but1);

        trans_button = new JButton(guistrings.getString("Run"));
        trans_button.setDefaultCapable(true);

        trans_button.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                Thread t = new Thread() {
                    @Override
                    public void run() {
                        trans_button.setEnabled(false);
                        PatternListWindow pw = null;
                        try {
                            trans.initTransitioner();
                            GeneratorTarget pwot = null;
                            if (pl != null) {
                                pl.clearList();
                                pwot = new GeneratorTarget(pl);
                                //jtp.setSelectedComponent(pl);
                            } else {
                                String title = trans.getNotationName() + " " + guistrings.getString("Patterns");
                                pw = new PatternListWindow(title, this);
                                pwot = new GeneratorTarget(pw);
                            }
                            trans.runTransitioner(pwot, max_patterns, max_time);
                            if (pl != null)
                                jtp.setSelectedComponent(pl);
                        } catch (JuggleExceptionDone ex) {
                            if (pl != null)
                                jtp.setSelectedComponent(pl);
                            Component parent = pw;
                            if (pw == null)
                                parent = pl;
                            new LabelDialog(parent, guistrings.getString("Generator_stopped_title"), ex.getMessage());
                        } catch (JuggleExceptionInterrupted jei) {
                            //System.out.println("generator thread quit");
                        } catch (JuggleExceptionUser ex) {
                            if (pw != null)
                                pw.dispose();
                            new ErrorDialog(ApplicationPanel.this, ex.getMessage());
                        } catch (Exception e) {
                            if (pw != null)
                                pw.dispose();
                            ErrorDialog.handleFatalException(e);
                        }

                        trans_button.setEnabled(true);
                    }
                };
                t.start();
            }
        });
        p2.add(trans_button);

        JPanel p3 = new JPanel();
        p3.setLayout(new BorderLayout());
        p3.add(p2, BorderLayout.LINE_END);

        p1.add(p3, BorderLayout.PAGE_END);

        jtp.addTab(guistrings.getString("Transitions"), p1);
    }

    protected void addGeneratorControl(Generator gen, PatternListPanel pl) {
        JPanel p1 = new JPanel();
        p1.setLayout(new BorderLayout());
        p1.add(gen.getGeneratorControl(), BorderLayout.PAGE_START);
        JPanel p2 = new JPanel();
        p2.setLayout(new FlowLayout(FlowLayout.TRAILING));

        JButton but1 = new JButton(guistrings.getString("Defaults"));

        but1.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                gen.resetGeneratorControl();
            }
        });
        p2.add(but1);

        gen_button = new JButton(guistrings.getString("Run"));

        gen_button.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                Thread t = new Thread() {
                    @Override
                    public void run() {
                        gen_busy.setVisible(true);
                        gen_button.setEnabled(false);
                        PatternListWindow pw = null;
                        try {
                            gen.initGenerator();
                            GeneratorTarget pwot = null;
                            if (pl != null) {
                                pl.clearList();
                                pwot = new GeneratorTarget(pl);
                                //jtp.setSelectedComponent(pl);
                            } else {
                                String title = gen.getNotationName() + " " + guistrings.getString("Patterns");
                                pw = new PatternListWindow(title, this);
                                pwot = new GeneratorTarget(pw);
                            }
                            gen.runGenerator(pwot, max_patterns, max_time);
                            if (pl != null)
                                jtp.setSelectedComponent(pl);
                        } catch (JuggleExceptionDone ex) {
                            if (pl != null)
                                jtp.setSelectedComponent(pl);
                            Component parent = pw;
                            if (pw == null)
                                parent = pl;
                            new LabelDialog(parent, guistrings.getString("Generator_stopped_title"), ex.getMessage());
                        } catch (JuggleExceptionInterrupted jei) {
                            //System.out.println("generator thread quit");
                        } catch (JuggleExceptionUser ex) {
                            if (pw != null)
                                pw.dispose();
                            new ErrorDialog(ApplicationPanel.this, ex.getMessage());
                        } catch (Exception e) {
                            if (pw != null)
                                pw.dispose();
                            ErrorDialog.handleFatalException(e);
                        }

                        gen_busy.setVisible(false);
                        gen_button.setEnabled(true);
                    }
                };
                t.start();
            }
        });
        p2.add(gen_button);

        gen_busy = new JLabel(guistrings.getString("Processing"));
        gen_busy.setVisible(false);
        JPanel p3 = new JPanel();
        p3.setLayout(new BorderLayout());
        JPanel p4 = new JPanel();
        GridBagLayout gb = new GridBagLayout();
        p4.setLayout(gb);
        p4.add(gen_busy);
        gb.setConstraints(gen_busy, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 0, new Insets(0,10,0,0)));
        p3.add(p4, BorderLayout.LINE_START);
        p3.add(p2, BorderLayout.LINE_END);

        p1.add(p3, BorderLayout.PAGE_END);

        jtp.addTab(guistrings.getString("Generator"), p1);
    }

    protected JButton getDefaultButton() {
        if (jtp == null)
            return null;
        if (jtp.getSelectedIndex() == 0)
            return juggle_button;
        else if (jtp.getSelectedIndex() == 1)
            return trans_button;
        else
            return gen_button;
    }
}

File: src/main/kotlin/jugglinglab/core/ApplicationPanel.kt
//
// ApplicationPanel.kt
//
// This class represents the entire contents of the ApplicationWindow frame.
// For a given notation type it creates a tabbed pane with separate panels for
// pattern entry, transitions, and generator.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.core

import jugglinglab.JugglingLab.guistrings
import jugglinglab.core.PatternWindow.Companion.bringToFront
import jugglinglab.generator.Generator
import jugglinglab.generator.Generator.Companion.newGenerator
import jugglinglab.generator.GeneratorTarget
import jugglinglab.generator.Transitioner
import jugglinglab.generator.Transitioner.Companion.newTransitioner
import jugglinglab.jml.JMLPattern.Companion.fromBasePattern
import jugglinglab.notation.NotationControl
import jugglinglab.notation.Pattern
import jugglinglab.notation.SiteswapNotationControl
import jugglinglab.util.*
import jugglinglab.util.ErrorDialog.handleFatalException
import jugglinglab.util.ErrorDialog.handleUserException
import jugglinglab.view.View
import java.awt.*
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.*
import javax.swing.event.ChangeEvent

class ApplicationPanel
    (
    private val parentFrame: JFrame?,
    // fields below are currently unused; they supported the applet version
    // of Juggling Lab
    private val animtarget: View?,
    private val patlist: PatternListPanel?
) : JPanel(), ActionListener {
    private var jtp: JTabbedPane? = null
    private var patlisttab: Boolean = false

    private var currentnum: Int = -1
    private var juggleButton: JButton? = null
    private var transButton: JButton? = null
    private var genButton: JButton? = null
    private var genBusy: JLabel? = null

    constructor(parentFrame: JFrame?) :
        this(parentFrame, animtarget = null, patlist = null)

    override fun actionPerformed(ae: ActionEvent) {
        val command = ae.getActionCommand()
        try {
            if (command.startsWith("notation")) {
                try {
                    val num = command.substring(8).toInt()
                    if (num != currentnum) {
                        setNotation(num)
                        parentFrame?.pack()
                        currentnum = num
                    }
                } catch (_: NumberFormatException) {
                    throw JuggleExceptionInternal("Error in notation number coding")
                }
            }
        } catch (e: Exception) {
            handleFatalException(e)
        }
    }

    // Input is for example Pattern.NOTATION_SITESWAP.

    fun setNotation(num: Int) {
        if (num > Pattern.builtinNotations.size) return
        if (jtp != null) {
            remove(jtp)
        }
        jtp = JTabbedPane()

        when (num) {
            Pattern.NOTATION_SITESWAP -> addPatternEntryControl(SiteswapNotationControl())
        }

        var pl = patlist
        if (pl == null && patlisttab) {
            pl = PatternListPanel(animtarget)
        }

        val trans = newTransitioner(Pattern.builtinNotations[num - 1])
        if (trans != null) {
            addTransitionerControl(trans, pl)
        }

        val gen = newGenerator(Pattern.builtinNotations[num - 1])
        if (gen != null) {
            addGeneratorControl(gen, pl)
        }

        if (pl != null) {
            jtp!!.addTab(guistrings.getString("Pattern_list_tab"), pl)
            if (patlist != null) {
                jtp!!.setSelectedComponent(pl) // if we loaded from a file
            }
        }

        // change the default button when the tab changes
        jtp!!.addChangeListener { _: ChangeEvent? -> rootPane.defaultButton = defaultButton }

        setLayout(BorderLayout())
        add(jtp!!, BorderLayout.CENTER)
        parentFrame?.rootPane?.defaultButton = defaultButton
    }

    private fun addPatternEntryControl(control: NotationControl) {
        val nbut1 = JButton(guistrings.getString("Defaults")).apply {
            addActionListener { _: ActionEvent? ->
                try {
                    control.resetControl()
                } catch (e: Exception) {
                    handleFatalException(e)
                }
            }
        }

        juggleButton =
            JButton(guistrings.getString("Juggle")).apply {
                setDefaultCapable(true)
                addActionListener { _: ActionEvent? ->
                    try {
                        val pl = control.parameterList
                        val p = control.newPattern().fromParameters(pl)
                        val jc = (AnimationPrefs()).fromParameters(pl)
                        pl.errorIfParametersLeft()

                        val notation = p.notationName
                        val config: String = p.toString()
                        val pat = fromBasePattern(notation, config)
                        pat.layoutPattern()

                        if (bringToFront(pat.hashCode)) {
                            return@addActionListener
                        }

                        if (animtarget != null) {
                            animtarget.restartView(pat, jc)
                        } else {
                            PatternWindow(pat.title, pat, jc)
                        }
                    } catch (je: JuggleExceptionUser) {
                        handleUserException(this@ApplicationPanel, je.message)
                    } catch (e: Exception) {
                        handleFatalException(e)
                    }
                }
            }

        val np2 = JPanel().apply {
            setLayout(FlowLayout(FlowLayout.TRAILING))
            add(nbut1)
            add(juggleButton)
        }
        val np1 = JPanel().apply {
            setLayout(BorderLayout())
            add(control, BorderLayout.PAGE_START)
            add(np2, BorderLayout.PAGE_END)
        }

        jtp!!.addTab(guistrings.getString("Pattern_entry"), np1)
    }

    private fun addTransitionerControl(trans: Transitioner, plp: PatternListPanel?) {
        val but1 = JButton(guistrings.getString("Defaults")).apply {
            addActionListener { _: ActionEvent? -> trans.resetTransitionerControl() }
        }

        transButton = JButton(guistrings.getString("Run")).apply {
            setDefaultCapable(true)
            addActionListener {
                val t: Thread =
                    object : Thread() {
                        override fun run() {
                            transButton!!.setEnabled(false)
                            var pw: PatternListWindow? = null
                            try {
                                trans.initTransitioner()
                                val pwot: GeneratorTarget?
                                if (plp != null) {
                                    plp.clearList()
                                    pwot = GeneratorTarget(plp)
                                    // jtp.setSelectedComponent(plp);
                                } else {
                                    val title =
                                        trans.notationName + " " + guistrings.getString("Patterns")
                                    pw = PatternListWindow(title, this)
                                    pwot = GeneratorTarget(pw.patternListPanel)
                                }
                                trans.runTransitioner(pwot, MAX_PATTERNS, MAX_TIME)
                                if (plp != null) {
                                    jtp!!.setSelectedComponent(plp)
                                }
                            } catch (ex: JuggleExceptionDone) {
                                if (plp != null) {
                                    jtp!!.setSelectedComponent(plp)
                                }
                                val parentComponent = pw ?: plp
                                LabelDialog(
                                    parentComponent,
                                    guistrings.getString("Generator_stopped_title"),
                                    ex.message
                                )
                            } catch (_: JuggleExceptionInterrupted) {
                                // System.out.println("generator thread quit");
                            } catch (ex: JuggleExceptionUser) {
                                pw?.dispose()
                                handleUserException(this@ApplicationPanel, ex.message)
                            } catch (e: Exception) {
                                pw?.dispose()
                                handleFatalException(e)
                            }

                            transButton!!.setEnabled(true)
                        }
                    }
                t.start()
            }
        }

        val p2 = JPanel().apply {
            setLayout(FlowLayout(FlowLayout.TRAILING))
            add(but1)
            add(transButton)
        }
        val p3 = JPanel().apply {
            setLayout(BorderLayout())
            add(p2, BorderLayout.LINE_END)
        }
        val p1 = JPanel().apply {
            setLayout(BorderLayout())
            add(trans.transitionerControl, BorderLayout.PAGE_START)
            add(p3, BorderLayout.PAGE_END)
        }

        jtp!!.addTab(guistrings.getString("Transitions"), p1)
    }

    private fun addGeneratorControl(gen: Generator, plp: PatternListPanel?) {
        val but1 = JButton(guistrings.getString("Defaults")).apply {
            addActionListener { _: ActionEvent? -> gen.resetGeneratorControl() }
        }

        genButton = JButton(guistrings.getString("Run")).apply {
            addActionListener {
                val t: Thread =
                    object : Thread() {
                        override fun run() {
                            genBusy!!.isVisible = true
                            genButton!!.setEnabled(false)
                            var pw: PatternListWindow? = null
                            try {
                                gen.initGenerator()
                                val pwot: GeneratorTarget?
                                if (plp != null) {
                                    plp.clearList()
                                    pwot = GeneratorTarget(plp)
                                    // jtp.setSelectedComponent(plp);
                                } else {
                                    val title =
                                        gen.notationName + " " + guistrings.getString("Patterns")
                                    pw = PatternListWindow(title, this)
                                    pwot = GeneratorTarget(pw.patternListPanel)
                                }
                                gen.runGenerator(pwot, MAX_PATTERNS, MAX_TIME)
                                if (plp != null) {
                                    jtp!!.setSelectedComponent(plp)
                                }
                            } catch (ex: JuggleExceptionDone) {
                                if (plp != null) {
                                    jtp!!.setSelectedComponent(plp)
                                }
                                val parentComponent = pw ?: plp
                                LabelDialog(
                                    parentComponent,
                                    guistrings.getString("Generator_stopped_title"),
                                    ex.message
                                )
                            } catch (_: JuggleExceptionInterrupted) {
                                // System.out.println("generator thread quit");
                            } catch (ex: JuggleExceptionUser) {
                                pw?.dispose()
                                handleUserException(this@ApplicationPanel, ex.message)
                            } catch (e: Exception) {
                                pw?.dispose()
                                handleFatalException(e)
                            }

                            genBusy!!.isVisible = false
                            genButton!!.setEnabled(true)
                        }
                    }
                t.start()
            }
        }

        genBusy = JLabel(guistrings.getString("Processing")).apply {
            isVisible = false
        }

        val gb = GridBagLayout().apply {
            setConstraints(
                genBusy, constraints(GridBagConstraints.LINE_START, 0, 0, Insets(0, 10, 0, 0))
            )
        }
        val p4 = JPanel().apply {
            setLayout(gb)
            add(genBusy)
        }
        val p2 = JPanel().apply {
            setLayout(FlowLayout(FlowLayout.TRAILING))
            add(but1)
            add(genButton)
        }
        val p3 = JPanel().apply {
            setLayout(BorderLayout())
            add(p4, BorderLayout.LINE_START)
            add(p2, BorderLayout.LINE_END)
        }
        val p1 = JPanel().apply {
            setLayout(BorderLayout())
            add(gen.generatorControl, BorderLayout.PAGE_START)
            add(p3, BorderLayout.PAGE_END)
        }

        jtp!!.addTab(guistrings.getString("Generator"), p1)
    }

    private val defaultButton: JButton?
        get() {
            if (jtp == null) {
                return null
            }
            return when (jtp!!.selectedIndex) {
                0 -> juggleButton
                1 -> transButton
                else -> genButton
            }
        }

    companion object {
        // execution limits for generator
        private const val MAX_PATTERNS: Int = 1000
        private const val MAX_TIME: Double = 15.0
    }
}
